File: kotlin/src/test/kotlin/com/bhengubv/circleai/brain/FusedRecallTest.kt
// FusedRecallTest.kt
//
// Verifies FusedRecall: Reciprocal Rank Fusion order, cross-source reinforcement,
// cold-start degradation to episodic, the graph confidence gate, empty-query
// short-circuit, and dedup by normalised text. Mirrors the TS pilot
// (tests/fused_recall.test.ts) and Go port.

package com.bhengubv.circleai.brain

import com.bhengubv.circleai.memory.brain.EpisodicEntry
import com.bhengubv.circleai.memory.brain.FusedRecall
import com.bhengubv.circleai.memory.brain.IEpisodicStore
import com.bhengubv.circleai.memory.brain.IHippoRagStore
import com.bhengubv.circleai.memory.brain.MemoryHit
import com.bhengubv.circleai.memory.brain.MemoryItem
import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.Test
import java.time.Instant
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class FusedRecallTest {

    // ── Test doubles ────────────────────────────────────────────────────────────

    private fun ep(id: String, userText: String): EpisodicEntry = EpisodicEntry(
        id = id,
        userText = userText,
        assistantText = "",
        recordedAtUtc = Instant.parse("2026-01-01T00:00:00Z"),
    )

    /** Episodic store that returns a fixed, pre-ranked list from searchAsync. */
    private class FakeEpisodic(private val hits: List<EpisodicEntry>) : IEpisodicStore {
        override suspend fun addAsync(entry: EpisodicEntry) {}
        override suspend fun searchAsync(queryEmbedding: FloatArray?, topK: Int): List<EpisodicEntry> =
            hits.take(topK)
        override suspend fun getRecentAsync(count: Int): List<EpisodicEntry> = hits.take(count)
        override suspend fun countAsync(): Int = hits.size
        override suspend fun pruneOlderThanAsync(cutoff: Instant): Int = 0
    }

    /** HippoRAG store that returns a fixed, pre-ranked list from multiHopRecallAsync. */
    private class FakeHippo(private val hits: List<MemoryHit>) : IHippoRagStore {
        override val backendId: String = "fake-hippo"
        override suspend fun indexAsync(item: MemoryItem) {}
        override suspend fun multiHopRecallAsync(query: String, topK: Int): List<MemoryHit> =
            hits.take(topK)
    }

    private fun graphHit(id: String, text: String, confidence: String? = null): MemoryHit {
        val metadata = if (confidence == null) null else mapOf("confidence" to confidence)
        return MemoryHit(MemoryItem(id, text, metadata), 1f)
    }

    // ── RRF ordering ────────────────────────────────────────────────────────────

    @Test
    fun `a memory surfaced by BOTH sources outranks one surfaced by only one`() = runTest {
        val episodic = FakeEpisodic(listOf(ep("a", "A"), ep("b", "B"), ep("c", "C")))
        val graph = FakeHippo(listOf(graphHit("g", "B"))) // reinforces B
        val recall = FusedRecall(episodic, graph)

        val hits = recall.recallAsync("q", null, 5)
        assertEquals(listOf("B", "A", "C"), hits.map { it.item.text })
    }

    @Test
    fun `cold-start (no graph) yields the episodic order unchanged`() = runTest {
        val episodic = FakeEpisodic(listOf(ep("a", "A"), ep("b", "B"), ep("c", "C")))
        val recall = FusedRecall(episodic, null)

        val hits = recall.recallAsync("q", null, 5)
        assertEquals(listOf("A", "B", "C"), hits.map { it.item.text })
    }

    @Test
    fun `respects topK`() = runTest {
        val episodic = FakeEpisodic(listOf(ep("a", "A"), ep("b", "B"), ep("c", "C")))
        val recall = FusedRecall(episodic, null)

        val hits = recall.recallAsync("q", null, 2)
        assertEquals(2, hits.size)
        assertEquals(listOf("A", "B"), hits.map { it.item.text })
    }

    // ── integrity gates ─────────────────────────────────────────────────────────

    @Test
    fun `drops graph hits below the confidence threshold`() = runTest {
        val episodic = FakeEpisodic(emptyList())
        val graph = FakeHippo(listOf(graphHit("low", "LOW", "0.2"), graphHit("high", "HIGH", "0.9")))
        val recall = FusedRecall(episodic, graph)

        val hits = recall.recallAsync("q", null, 5)
        val texts = hits.map { it.item.text }
        assertFalse(texts.contains("LOW"), "below-threshold hit must be dropped")
        assertTrue(texts.contains("HIGH"))
    }

    @Test
    fun `keeps graph hits that carry no confidence metadata (gate is a no-op)`() = runTest {
        val episodic = FakeEpisodic(emptyList())
        val graph = FakeHippo(listOf(graphHit("g", "NOCONF")))
        val recall = FusedRecall(episodic, graph)

        val hits = recall.recallAsync("q", null, 5)
        assertEquals(listOf("NOCONF"), hits.map { it.item.text })
    }

    @Test
    fun `skips the graph entirely for an empty or whitespace query`() = runTest {
        val episodic = FakeEpisodic(listOf(ep("a", "A")))
        val graph = FakeHippo(listOf(graphHit("g", "GRAPH")))
        val recall = FusedRecall(episodic, graph)

        val hits = recall.recallAsync("   ", null, 5)
        val texts = hits.map { it.item.text }
        assertEquals(listOf("A"), texts)
        assertFalse(texts.contains("GRAPH"))
    }

    @Test
    fun `degrades to episodic when the graph throws`() = runTest {
        val episodic = FakeEpisodic(listOf(ep("a", "A")))
        val throwing = object : IHippoRagStore {
            override val backendId: String = "boom"
            override suspend fun indexAsync(item: MemoryItem) {}
            override suspend fun multiHopRecallAsync(query: String, topK: Int): List<MemoryHit> =
                throw RuntimeException("graph unavailable")
        }
        val recall = FusedRecall(episodic, throwing)

        val hits = recall.recallAsync("q", null, 5)
        assertEquals(listOf("A"), hits.map { it.item.text })
    }

    // ── dedup ───────────────────────────────────────────────────────────────────

    @Test
    fun `fuses two hits with the same normalised text into one entry`() = runTest {
        val episodic = FakeEpisodic(listOf(ep("a", "Durban  Weather")))
        val graph = FakeHippo(listOf(graphHit("g", "durban weather"))) // same key
        val recall = FusedRecall(episodic, graph)

        val hits = recall.recallAsync("q", null, 5)
        assertEquals(1, hits.size)
    }
}

File: rust/tests/fused_recall_test.rs
//! fused_recall_test.rs
//!
//! Verifies FusedRecall: Reciprocal Rank Fusion order, cross-source
//! reinforcement, cold-start degradation to episodic, the graph confidence gate,
//! empty-query short-circuit, and dedup by normalised text. Mirrors the TS pilot
//! suite tests/fused_recall.test.ts and the Go suite fused_recall_test.go 1:1.

use std::collections::HashMap;
use std::sync::Arc;

use chrono::{TimeZone, Utc};
use circle_ai::brain::BrainError;
use circle_ai::memory::episodic::EpisodicSearch;
use circle_ai::memory::graph::{IHippoRagStore, MemoryHit, MemoryItem};
use circle_ai::memory::recall::{FusedRecall, IRecall};
use circle_ai::memory::EpisodicMemoryEntry;
use uuid::Uuid;

// ── Test doubles ─────────────────────────────────────────────────────────────

fn ep_entry(_id: &str, user_text: &str) -> EpisodicMemoryEntry {
    // The fused-recall suite keys on text, never on id; a fresh v4 id suffices.
    EpisodicMemoryEntry {
        id: Uuid::new_v4(),
        recorded_at_utc: Utc.with_ymd_and_hms(2026, 1, 1, 0, 0, 0).unwrap(),
        user_text: user_text.to_string(),
        assistant_text: String::new(),
        app_context: None,
        embedding: None,
        tags: None,
    }
}

/// Returns a fixed, pre-ranked list from `search`.
struct FakeEpisodic {
    hits: Vec<EpisodicMemoryEntry>,
}

impl EpisodicSearch for FakeEpisodic {
    fn search(
        &self,
        _query_embedding: Option<&[f32]>,
        top_k: usize,
    ) -> Result<Vec<EpisodicMemoryEntry>, BrainError> {
        Ok(self.hits.iter().take(top_k).cloned().collect())
    }
}

/// Returns a fixed, pre-ranked list from `multi_hop_recall`.
struct FakeHippo {
    hits: Vec<MemoryHit>,
}

impl IHippoRagStore for FakeHippo {
    fn backend_id(&self) -> &str {
        "fake-hippo"
    }
    fn index(&self, _item: &MemoryItem) -> Result<(), BrainError> {
        Ok(())
    }
    fn multi_hop_recall(&self, _query: &str, top_k: usize) -> Result<Vec<MemoryHit>, BrainError> {
        Ok(self.hits.iter().take(top_k).cloned().collect())
    }
}

/// Always errors from `multi_hop_recall`.
struct ThrowingHippo;

impl IHippoRagStore for ThrowingHippo {
    fn backend_id(&self) -> &str {
        "boom"
    }
    fn index(&self, _item: &MemoryItem) -> Result<(), BrainError> {
        Ok(())
    }
    fn multi_hop_recall(&self, _query: &str, _top_k: usize) -> Result<Vec<MemoryHit>, BrainError> {
        Err(BrainError::new("graph unavailable"))
    }
}

fn graph_hit(id: &str, text: &str, confidence: Option<&str>) -> MemoryHit {
    let metadata = confidence.map(|c| {
        let mut m = HashMap::new();
        m.insert("confidence".to_string(), c.to_string());
        m
    });
    MemoryHit {
        item: MemoryItem {
            id: id.to_string(),
            text: text.to_string(),
            metadata,
        },
        score: 1.0,
    }
}

fn hit_texts(hits: &[MemoryHit]) -> Vec<String> {
    hits.iter().map(|h| h.item.text.clone()).collect()
}

fn must_recall(
    ep: Arc<dyn EpisodicSearch>,
    g: Option<Arc<dyn IHippoRagStore>>,
) -> FusedRecall {
    FusedRecall::new(ep, g, None).expect("NewFusedRecall")
}

// ── RRF ordering ─────────────────────────────────────────────────────────────

#[test]
fn a_memory_surfaced_by_both_sources_outranks_one_surfaced_by_only_one() {
    let episodic = Arc::new(FakeEpisodic {
        hits: vec![ep_entry("a", "A"), ep_entry("b", "B"), ep_entry("c", "C")],
    });
    let graph = Arc::new(FakeHippo {
        hits: vec![graph_hit("g", "B", None)], // reinforces B
    });
    let recall = must_recall(episodic, Some(graph));

    let hits = recall.recall("q", None, 5).expect("Recall");
    assert_eq!(hit_texts(&hits), vec!["B", "A", "C"]);
}

#[test]
fn cold_start_no_graph_yields_the_episodic_order_unchanged() {
    let episodic = Arc::new(FakeEpisodic {
        hits: vec![ep_entry("a", "A"), ep_entry("b", "B"), ep_entry("c", "C")],
    });
    let recall = must_recall(episodic, None);

    let hits = recall.recall("q", None, 5).expect("Recall");
    assert_eq!(hit_texts(&hits), vec!["A", "B", "C"]);
}

#[test]
fn rrf_respects_top_k() {
    let episodic = Arc::new(FakeEpisodic {
        hits: vec![ep_entry("a", "A"), ep_entry("b", "B"), ep_entry("c", "C")],
    });
    let recall = must_recall(episodic, None);

    let hits = recall.recall("q", None, 2).expect("Recall");
    assert_eq!(hits.len(), 2);
    assert_eq!(hit_texts(&hits), vec!["A", "B"]);
}

// ── Integrity gates ──────────────────────────────────────────────────────────

#[test]
fn drops_graph_hits_below_the_confidence_threshold() {
    let episodic = Arc::new(FakeEpisodic { hits: vec![] });
    let graph = Arc::new(FakeHippo {
        hits: vec![
            graph_hit("low", "LOW", Some("0.2")),
            graph_hit("high", "HIGH", Some("0.9")),
        ],
    });
    let recall = must_recall(episodic, Some(graph));

    let hits = recall.recall("q", None, 5).expect("Recall");
    let texts = hit_texts(&hits);
    assert!(!texts.contains(&"LOW".to_string()), "below-threshold hit must be dropped");
    assert!(texts.contains(&"HIGH".to_string()), "HIGH must be kept");
}

#[test]
fn keeps_graph_hits_that_carry_no_confidence_metadata() {
    let episodic = Arc::new(FakeEpisodic { hits: vec![] });
    let graph = Arc::new(FakeHippo {
        hits: vec![graph_hit("g", "NOCONF", None)],
    });
    let recall = must_recall(episodic, Some(graph));

    let hits = recall.recall("q", None, 5).expect("Recall");
    assert_eq!(hit_texts(&hits), vec!["NOCONF"]);
}

#[test]
fn skips_the_graph_entirely_for_an_empty_query() {
    let episodic = Arc::new(FakeEpisodic {
        hits: vec![ep_entry("a", "A")],
    });
    let graph = Arc::new(FakeHippo {
        hits: vec![graph_hit("g", "GRAPH", None)],
    });
    let recall = must_recall(episodic, Some(graph));

    let hits = recall.recall("   ", None, 5).expect("Recall");
    let texts = hit_texts(&hits);
    assert_eq!(texts, vec!["A"]);
    assert!(!texts.contains(&"GRAPH".to_string()), "graph must be skipped for empty query");
}

#[test]
fn degrades_to_episodic_when_the_graph_errors() {
    let episodic = Arc::new(FakeEpisodic {
        hits: vec![ep_entry("a", "A")],
    });
    let recall = must_recall(episodic, Some(Arc::new(ThrowingHippo)));

    let hits = recall.recall("q", None, 5).expect("Recall");
    assert_eq!(hit_texts(&hits), vec!["A"]);
}

// ── Dedup ────────────────────────────────────────────────────────────────────

#[test]
fn fuses_two_hits_with_the_same_normalised_text_into_one_entry() {
    let episodic = Arc::new(FakeEpisodic {
        hits: vec![ep_entry("a", "Durban  Weather")],
    });
    let graph = Arc::new(FakeHippo {
        hits: vec![graph_hit("g", "durban weather", None)], // same key
    });
    let recall = must_recall(episodic, Some(graph));

    let hits = recall.recall("q", None, 5).expect("Recall");
    assert_eq!(hits.len(), 1);
}
